### PR TITLE
[CCB] | MultiClientTransactionStarter refactor - part 1

### PR DESCRIPTION
**Pull request by sudiksha27**
**Goals (and why)**:
Refactor TransactionStarter to accommodate the flow for batching startTransactions requests across multiple clients.

**Implementation Description (bullets)**:

- Extract out IdentifiedAtlasDbTransactionStarter interface, this will have 2 implementations - 1. batching requests per client, 2. batching startTransactions requests across clients
- TransactionStarter now delegates startTransactions calls to IdentifiedAtlasDbTransactionStarter
- Refactor out startTransactions batching logic from TransactionStarter to BatchingIdentifiedAtlasDbTransactionStarter.java
- Move startTransactions utility methods from TransactionStarter to TransactionStarterHelper

**Testing (What was existing testing like?  What have you done to improve it?)**:
Existing tests should suffice

**Concerns (what feedback would you like?)**:
Does this refactor break something?

**Where should we start reviewing?**:

- BatchingIdentifiedAtlasDbTransactionStarter

**Priority (whenever / two weeks / yesterday)**:
Before EOD tomorrow.

### CODE REVIEW

File: lock-api/src/main/java/com/palantir/lock/client/BatchingIdentifiedAtlasDbTransactionStarter.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import com.palantir.lock.watch.LockWatchEventCache;
+import com.palantir.lock.watch.LockWatchVersion;
+import com.palantir.logsafe.Preconditions;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+public class BatchingIdentifiedAtlasDbTransactionStarter implements IdentifiedAtlasDbTransactionStarter {
```
[the remark this answers is not in the document]
Comment (Jolyon-S) [reply lock-api/src/main/java/com/palantir/lock/client/BatchingIdentifiedAtlasDbTransactionStarter.java:37]: [Sanity check]: This class has essentially just extracted the logic that used to live in TransactionStarter, right?

File: lock-api/src/main/java/com/palantir/lock/client/TransactionStarterHelper.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+import com.palantir.lock.v2.PartitionedTimestamps;
+import com.palantir.lock.v2.StartIdentifiedAtlasDbTransactionResponse;
+import com.palantir.lock.v2.TimestampAndPartition;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public final class TransactionStarterHelper {
```
Comment (Jolyon-S) [inline lock-api/src/main/java/com/palantir/lock/client/TransactionStarterHelper.java:34]: I'm not sure about this class. Why did you need to extract these methods out? I understand delegating to the different start transaction implementations, but not sure why you need this helper (I could be missing a part of the grand design, ofc)

**Review by Jolyon-S — COMMENTED** [review]
Couple of nits, couple of questions

File: lock-api/src/main/java/com/palantir/lock/client/TransactionStarterHelper.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    private TransactionStarterHelper() {
+        // Do not instantiate helper class
+    }
+
+    static Set<LockToken> unlock(Set<LockToken> tokens, LockLeaseService lockLeaseService) {
```
Comment (jeremyk-91) [inline lock-api/src/main/java/com/palantir/lock/client/TransactionStarterHelper.java:39]: This is pretty intense but is just extracted from the original.

**Review by jeremyk-91 — APPROVED** [review]
🕐 0:15 🏁 2:17

👍  one thing, but generally looks good.

File: lock-api/src/main/java/com/palantir/lock/client/TransactionStarterHelper.java (hunk 2 above)
Comment (sudiksha27) [reply lock-api/src/main/java/com/palantir/lock/client/TransactionStarterHelper.java:34]: Ah, right. There is a draft here - https://github.com/palantir/atlasdb/pull/5270 for when we have the multiClient batching which needs access to these utility methods.

File: lock-api/src/main/java/com/palantir/lock/client/BatchingIdentifiedAtlasDbTransactionStarter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sudiksha27) [reply lock-api/src/main/java/com/palantir/lock/client/BatchingIdentifiedAtlasDbTransactionStarter.java:37]: That's right.

**Comment by Jolyon-S** [comment]
Fine from my end; let's go!
